Add initial 'weather' mod to vary cloud density, thickness, velocity

paramat opened the pull request:
![screenshot_20191001_171356](https://user-images.githubusercontent.com/3686677/65981298-b4111800-e470-11e9-9829-e3c754ee3d2d.png)

![screenshot_20191001_172037](https://user-images.githubusercontent.com/3686677/65981309-b83d3580-e470-11e9-8763-b5f238ae2cb2.png)

Firstly, don't get too excited by the mod name.
This mod is like 'env_sounds', it is intended to be simple at first, to have more features added later as engine features improve. This is why it is called 'weather' even though it only varies cloud density, thickness and velocity.

The reason i haven't included dark clouds, rain. snow etc. yet is because those are currently problematic/intensive/network-intensive. Work on improving weather implementation is going on elsewhere (Hidden Worlds game, snowdrift mod). Considering the problems it is too early to add a complete weather system to MTG, the MTEngine may need improving first.
Clouds are very often in view, so although this is simple it makes a very significant difference to the experience.
Using my experience with the snowdrift mod (and Hidden Worlds) i will be happy to develop this mod and add weather features in future.

* This mod uses player location humidity and 4 noises to vary: density, thickness, x velocity, z velocity.
* By default, every 8sec all players have their sky updated.
* The default time scale for changes is 600sec / 10min, half of a default MT day. This keeps changes gradual but allows significant change during daylight.
* Every server session start a random time offset is added to randomise the behaviour of the noises. Without this, each server session in a world would go through identical noise variations.
* Also works with 3D clouds disabled.

Variations
------------
Density (proportion of cloud covering sky):

The initial value varies from 0.25 at humidity 0 to 1.0 at humidity 100, and is limited to those values for more extreme humidities.
Then this value is multiplied by the noise value (0.0 to 1.0).
The result is that density varies between 0.0 and 0.25 at humidity 0 (desert), and varies between 0.0 and 1.0 at humidity 100 (rainforest).
The intention being that clear sky is possible in both biomes, but cloud cover is always low in desert, but can become overcast in rainforest.

Thickness (in nodes):

The initial value varies from 8 (fairly thin clouds) at humidity 0 to 32 (twice classic thickness) at humidity 100, and is limited to those values for more extreme humidities.
Then this value is multiplied by the noise value (0.0 to 1.0).
At this stage, the value varies between 0 and 8 at humidity 0 (desert), and varies between 0 and 32 at humidity 100 (rainforest).
The value is then rounded down to an integer to make clouds a certain number of nodes thick, in the same way that their unit width is 64 nodes.
Finally a minimum of 1 node thick is enforced, so that only density variaton can make clouds disappear, not thickness variation.
The intention being that thin clouds are possible in both biomes, but clouds are always fairly thin in desert, but can become very fat in rainforest.

Velocity (in nodes per second):

Each component is varied by noise between 0 and 4 (twice classic velocity).

How to test
--------------
* In the init.lua, set TSCALE = 3 and CYCLE = 3, to cause rapidly changing noise values.
* Edit line 78 to `local humid = 0` or `local humid = 100` to see behaviour at the extremes of humidity, indoependent of what biome the player is in.

Fixer-007 commented:
I guess it is a good start, you always start with something and build up it along the way.

Panquesito7 commented on mods/weather/init.lua:
```diff
@@ -0,0 +1,96 @@
+	local n_speedz = nobj_speedz:get2d({x = time, y = 0})
+
+	for _, player in ipairs(minetest.get_connected_players()) do
+		local humid = minetest.get_humidity(player:getpos())
```
Please use `get_pos`.
```suggestion
		local humid = minetest.get_humidity(player:get_pos())
```

paramat replied:
Thanks will do.

paramat replied:
Done.

Panquesito7 replied:
Thank you!

lhofhansl commented:
I like it. +1

Any possibility to generate the cloud with 3D noise, so that they have variations along the Y-axis too?

paramat commented:
Added code to disable the mod in mgv6 and mgsinglenode, where the Biome API doesn't function.
Added a setting to disable the mod, for people using alternative weather mods.

lhofhansl, only with engine changes.

Panquesito7 commented on mods/weather/init.lua:
```diff
@@ -0,0 +1,105 @@
+
+local mg_name = minetest.get_mapgen_setting("mg_name")
+if mg_name == "v6" or mg_name == "singlenode" or
+		minetest.settings:get_bool("enable_weather") == false then
```
```suggestion
		minetest.settings:get_bool("enable_weather") ~= true then
```

paramat replied:
This was intentional so that weather will be enabled if the setting is not defined (nil).

DevyHeavy commented:
I tested this. I found that when a world is first started, or a player joins, the clouds are not updated until after CYCLE time, as expected. However, it looks weird to have the clouds suddenly change since they are not updated right away using this mod.

paramat commented:
Yes good point, i should try to trigger an update on player join.
WIP

sfan5 commented on mods/weather/init.lua:
```diff
@@ -0,0 +1,112 @@
+			density = math.min(math.max(humid / 100, 0.25), 1.0) * n_density,
+			thickness = math.max(math.floor(
+				math.min(math.max(32 * humid / 100, 8), 32) * n_thickness
+				), 1),
```
This code is not very nice to read.
A helper `local clamp = function(lower, upper, n) return math.min(math.max(n, lower), upper)` can be used to improve this.

paramat replied:
I agree.

paramat replied:
Done.
